### ISSUE TITLE
Support breakpoints for responsive design

### ISSUE THREAD

**Issue by smlarkin** [vadimdemedes/tailwind-rn#50]
Understandably, on mobile, breakpoints are not super relevant.

However, RN can "go back to the web" using [react-native-web](https://github.com/necolas/react-native-web), which is now baked into [Expo managed workflows](https://docs.expo.io/workflow/web/).

Currently, you must combine a [design system](https://github.com/Shopify/restyle) or [library](https://github.com/wcandillon/react-native-responsive-ui) or ["roll your own"](https://github.com/nandorojo/dripsy#before-dripsy-%EF%B8%8F) to get responsive breakpoint support in RN.

Having breakpoints carry over to Tailwind-RN ([from vanilla Tailwind](https://tailwindcss.com/docs/responsive-design)) would be helpful for anyone trying to write an app in a single codebase, and it would avoid having to use more design systems and libraries on top of Tailwind.

**Comment by SleeplessByte** [comment]
Here is a way to do it without waiting on #56.

Uses [`@expo/match-media`](https://github.com/expo/match-media) and [`react-responsive`](https://github.com/contra/react-responsive).

```typescript
import { useMediaQuery } from 'react-responsive';

function useIsAtLeastSm() {
  return useMediaQuery({ minWidth: 640 });
}

function useIsAtLeastMd() {
  return useMediaQuery({ minWidth: 768 });
}

function useIsAtLeastLg() {
  return useMediaQuery({ minWidth: 1024 });
}

function useIsAtLeastXl() {
  return useMediaQuery({ minWidth: 1280 });
}

function tailwindResponsive(
  always: string,
  { sm, md, lg, xl }: { sm?: string; md?: string; lg?: string; xl?: string },
  {
    sm: applySm,
    md: applyMd,
    lg: applyLg,
    xl: applyXl,
  }: {
    sm?: boolean;
    md?: boolean;
    lg?: boolean;
    xl?: boolean;
  }
) {
  return tailwind(
    [always, applySm && sm, applyMd && md, applyLg && lg, applyXl && xl]
      .filter(Boolean)
      .join(' ')
  );
}
```

```tsx
function MyScreen() {
  const sm = useIsAtLeastSm();
  const md = useIsAtLeastMd();
  const xl = useIsAtLeastXl();
  
  return (
   <View
    style={tailwindResponsive(
      'mt-16 max-w-7xl px-4 self-center',
      { sm: 'mt-24' },
      { sm }
    )}
   >
    <View style={tailwind('self-center items-center')}>
      <View
        style={tailwindResponsive(
          'flex items-center',
          { xl: 'flex-row' },
          { xl }
        )}
      >
        <Text
          style={tailwindResponsive(
            'flex text-4xl tracking-tight font-extrabold text-gray-900 text-center',
            { sm: 'text-5xl', md: 'text-6xl', xl: 'mr-2' },
            { sm, md, xl }
          )}
        >
          Here is a way for
        </Text>
        <Text
          style={tailwindResponsive(
            'flex text-4xl tracking-tight font-extrabold text-indigo-600 text-center',
            { sm: 'text-5xl', md: 'text-6xl' },
            { sm, md }
          )}
        >
          responsive tailwind
        </Text>
      </View>
      <Paragraph
        style={tailwindResponsive(
          'mt-3 max-w-md text-base text-gray-500 text-center',
          { sm: 'text-lg', md: 'mt-5 text-xl max-w-3xl' },
          { sm, md }
        )}
      >
        Is it great? Eh. Is it easy. I think so. Why don't you call the
        hooks inside something like useTailwindResponsive? Because that
        would cause each useTailwindResponsive to have 4+ hooks running,
        when you only need one or two. They would also ALL subscribe to the
        media queries.
      </Paragraph>
    </View>
  </View>
)
```

**Comment by fkapsahili** [comment]
If this helps someone like me, here is a custom hook that can be easily reused in any component afterwards:

```javascript
export default function Example() {
  const { tailwindResponsive, sm, md, lg, xl } = useTailwind();
        <View
          style={tailwindResponsive(
            "mt-16 max-w-7xl px-4 self-center",
            { sm: "mt-24" },
            { sm }
          )}
        >
          <Text>Not much going on here</Text>
          <View style={tailwind("self-center items-center")}></View>
        </View>
   }
````

```javascript
import tailwind from "tailwind-rn";
import { useMediaQuery } from "react-responsive";

export default function useTailwind() {
  function useIsAtLeastSm() {
    return useMediaQuery({ minWidth: 640 });
  }

  function useIsAtLeastMd() {
    return useMediaQuery({ minWidth: 768 });
  }

  function useIsAtLeastLg() {
    return useMediaQuery({ minWidth: 1024 });
  }

  function useIsAtLeastXl() {
    return useMediaQuery({ minWidth: 1280 });
  }

  const sm = useIsAtLeastSm();
  const md = useIsAtLeastMd();
  const lg = useIsAtLeastLg();
  const xl = useIsAtLeastXl();

  function tailwindResponsive(
    always: string,
    { sm, md, lg, xl }: { sm?: string; md?: string; lg?: string; xl?: string },
    {
      sm: applySm,
      md: applyMd,
      lg: applyLg,
      xl: applyXl,
    }: {
      sm?: boolean;
      md?: boolean;
      lg?: boolean;
      xl?: boolean;
    }
  ) {
    return tailwind(
      [always, applySm && sm, applyMd && md, applyLg && lg, applyXl && xl]
        .filter(Boolean)
        .join(" ")
    );
  }

  return { tailwindResponsive, sm, md, lg, xl };
}
```

**Comment by nandorojo** [comment]
Here's an idea for a simpler API:

```
import { View } from 'tailwind-rn'

<View 
  tw="md:w-32"
/>
```

Under the hood, `View` can use `useWindowDimensions` to determine the responsive value/breakpoint. This is what I do for `dripsy`.

**Comment by vadimdemedes** [comment]
Hey folks, I know it's been a year, but I just released [v4.0.0](https://github.com/vadimdemedes/tailwind-rn/releases/tag/v4.0.0) with a built in support for responsive design and even custom `min-width`, `max-width`, `min-height` and `max-height` breakpoints!
